sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -256,13 +255,9 @@ class QueryExecution(
 
     // trigger to compute stats for logical plans
     try {
-      optimizedPlan.foreach(_.expressions.foreach(_.foreach {
-        case subqueryExpression: SubqueryExpression =>
-          // trigger subquery's child plan stats propagation
-          subqueryExpression.plan.stats
-        case _ =>
-      }))
-      optimizedPlan.stats
+      optimizedPlan.collectWithSubqueries {
+        case plan => plan.stats
```

Comment:
I see, so this basically triggers stats calculation for every node in the plan tree.

Reply:
Yes, right.

Reply:
Can we add a comment to explain why we need to do it? It's not that obvious to many reviewers.

Reply:
Yes, I added a comment.